runelite-client/src/main/java/net/runelite/client/plugins/multicombatborders/MultiCombatBordersPlugin.java

```diff
@@ -0,0 +1,140 @@
+import java.util.stream.Collectors;
+import javax.annotation.Nonnull;
+import javax.inject.Inject;
+import net.runelite.api.Client;
+import net.runelite.api.coords.WorldPoint;
+import net.runelite.api.events.GameTick;
+import net.runelite.client.config.ConfigManager;
+import net.runelite.client.plugins.Plugin;
+import net.runelite.client.plugins.PluginDescriptor;
+import net.runelite.client.plugins.opponentinfo.OpponentInfoPlugin;
+import net.runelite.client.ui.overlay.OverlayManager;
+
+@PluginDescriptor(
+	name = "Multi-combat borders",
+	description = "Show borders between mutli and single combat areas",
+	enabledByDefault = false
+)
+public class MultiCombatBordersPlugin extends Plugin
+{
+	@Inject
+	private Client client;
+
+	@Inject
+	private OverlayManager overlayManager;
+
+	@Inject
+	private MultiCombatBordersOverlay multiCombatBordersOverlay;
+
+	private List<MultiCombatBorder> multiCombatBorders = Collections.emptyList();
+	List<MultiCombatBorder> multiCombatBordersWithinView = Collections.emptyList();
+	private WorldPoint prevPlayerLocation;
+
+	@Provides
+	private MultiCombatBordersConfig getConfig(ConfigManager configManager)
+	{
+		return configManager.getConfig(MultiCombatBordersConfig.class);
+	}
+
+	@Override
+	protected void startUp() throws Exception
+	{
+		multiCombatBorders = loadMultiCombatBorders();
+		overlayManager.add(multiCombatBordersOverlay);
+	}
+
+	@Override
+	protected void shutDown() throws Exception
+	{
+		overlayManager.remove(multiCombatBordersOverlay);
+	}
+
+	private static List<MultiCombatBorder> loadMultiCombatBorders()
+	{
+		Gson gson = new Gson();
+		InputStream multiBordersFile = OpponentInfoPlugin.class.getResourceAsStream("/multi_borders.json");
+		JsonArray items = gson.fromJson(new InputStreamReader(multiBordersFile), JsonArray.class);
+
+		List<MultiCombatBorder> output = new ArrayList<>(items.size());
+
+		for (JsonElement element : items)
+		{
+			JsonObject item = (JsonObject) element;
+			for (String key : item.keySet())
+			{
+				JsonObject nested = item.getAsJsonObject(key);
+				output.add(
+					new MultiCombatBorder(
+						key,
+						fromJsonObject(nested.getAsJsonObject("single")),
+						fromJsonObject(nested.getAsJsonObject("multi"))));
+			}
+		}
+
+		return output;
+	}
+
+	private static WorldPoint fromJsonObject(@Nonnull JsonObject jsonObject)
+	{
+		JsonPrimitive x = jsonObject.getAsJsonPrimitive("x");
+		JsonPrimitive y = jsonObject.getAsJsonPrimitive("y");
+		return new WorldPoint(x.getAsInt(), y.getAsInt(), 0);
+	}
+
+	@Subscribe
+	void onTick(GameTick tick)
+	{
+		WorldPoint playerLocation = client.getLocalPlayer().getWorldLocation();
+		if (playerLocation != null && !playerLocation.equals(prevPlayerLocation))
+		{
+			prevPlayerLocation = playerLocation;
+			multiCombatBordersWithinView = getBordersWithinView(playerLocation);
+		}
+	}
+
+	private List<MultiCombatBorder> getBordersWithinView(@Nonnull WorldPoint playerLocation)
+	{
+		return multiCombatBorders
```

Comment:
It still filters through each border when you are moving though.

Reply:
Quickly tried a chunked solution, 8x8 tile chunks, loading 2 chunks in each direction (5x5)
Filtering dropped to 1/5th. On my machine from ~ 0.75 ms -> 0.15 ms.